flang/lib/Parser/openacc-parsers.cpp

```diff
@@ -221,11 +228,19 @@ TYPE_PARSER(sourced(construct<AccBeginBlockDirective>(
     sourced(Parser<AccBlockDirective>{}), Parser<AccClauseList>{})))
 
 TYPE_PARSER(startAccLine >> sourced(construct<AccEndBlockDirective>("END"_tok >>
-                                sourced(Parser<AccBlockDirective>{}))))
+                                sourced(recovery(Parser<AccBlockDirective>{},
+                                    construct<AccBlockDirective>(pure(
+                                        llvm::acc::Directive::ACCD_data)))))))
 
 TYPE_PARSER(construct<OpenACCBlockConstruct>(
     Parser<AccBeginBlockDirective>{} / endAccLine, block,
-    Parser<AccEndBlockDirective>{} / endAccLine))
+    // TODO: This still allows mismatched directives.
+    recovery(withMessage("expected OpenACC end block directive"_err_en_US,
+                 attempt(Parser<AccEndBlockDirective>{} / endAccLine)),
+        // TODO: Is there a simpler way to build this?
+        sourced(construct<AccEndBlockDirective>(
```

Comment:
Are you sure that you want `sourced` here?  It's going to create `AccEndBlockDirective` and `AccBlockDirective` class instances whose `source` data members will point to something that isn't such a thing, and that might cause semantic error messages on those parse tree nodes to point to arbitrary locations in the source file.

Reply:
No, I will remove the calls to source.